edge-util/src/Microsoft.Azure.Devices.Edge.Storage/DbStoreProviderWithBackupRestore.cs

```diff
@@ -0,0 +1,351 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Storage
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+    using Newtonsoft.Json;
+    using Newtonsoft.Json.Converters;
+
+    /// <summary>
+    /// Wraps backup and restore functionality around a DB Store Provider.
+    /// </summary>
+    public class DbStoreProviderWithBackupRestore : DbStoreProviderDecorator
+    {
+        const string BackupMetadataFileName = "meta.json";
+        const string DefaultStoreBackupName = "$Default";
+        readonly string backupPath;
+        readonly ISet<string> dbStores;
+        readonly IDbStoreBackupRestore dbStoreBackupRestore;
+        readonly SerializationFormat backupFormat;
+        readonly Events events;
+
+        DbStoreProviderWithBackupRestore(
+            IDbStoreProvider dbStoreProvider,
+            string backupPath,
+            IDbStoreBackupRestore dbStoreBackupRestore,
+            SerializationFormat backupFormat)
+            : base(dbStoreProvider)
+        {
+            this.backupPath = Preconditions.CheckNonWhiteSpace(backupPath, nameof(backupPath));
+            this.dbStoreBackupRestore = Preconditions.CheckNotNull(dbStoreBackupRestore, nameof(dbStoreBackupRestore));
+            this.backupFormat = backupFormat;
+            this.dbStores = new HashSet<string>();
+            this.events = new Events(this.Log);
+        }
+
+        public static async Task<IDbStoreProvider> CreateAsync(
+            IDbStoreProvider dbStoreProvider,
+            string backupPath,
+            IDbStoreBackupRestore dbStoreBackupRestore,
+            SerializationFormat backupFormat)
+        {
+            DbStoreProviderWithBackupRestore provider = new DbStoreProviderWithBackupRestore(dbStoreProvider, backupPath, dbStoreBackupRestore, backupFormat);
+            await provider.RestoreAsync();
+            return provider;
+        }
+
+        private async Task RestoreAsync()
+        {
+            string backupMetadataFilePath = Path.Combine(this.backupPath, BackupMetadataFileName);
+
+            if (!File.Exists(backupMetadataFilePath))
+            {
+                this.events.NoBackupsForRestore();
+                return;
+            }
+
+            try
+            {
+                string fileText = File.ReadAllText(backupMetadataFilePath);
+                BackupMetadataList backupMetadataList = JsonConvert.DeserializeObject<BackupMetadataList>(fileText);
```

Comment:
Code that is related to IO and deserialization I think it should be moved to dbStoreBackupStore

Reply:
DbStoreBackupRestore has been removed as per our conversation and all that code has been moved to DbStoreProviderWithBackupRestore